Route public replies to all actors instead of just followers

pfefferle opened the pull request:
## What does this PR do?

Routes public ActivityPub replies to followers of the reply author, while maintaining the ability to add additional explicitly-addressed recipients.

## What problem does it solve?

Fixes #2537
Related to #2539

When receiving public replies via ActivityPub, they were only being delivered to followers of the person replying. This implementation ensures followers receive the reply while still allowing the function to process additional recipients from the activity's addressing fields.

## How did you implement your changes?

Modified `get_local_recipients()` in `includes/rest/class-inbox-controller.php`:
- Initialize `$user_ids` array at the beginning of the method
- For public activities, populate `$user_ids` with followers of the actor
- Remove early return, allowing the function to continue processing
- The subsequent recipient extraction logic can add any additional explicitly-addressed recipients to the follower list
- This maintains backward compatibility while fixing the conversation continuity issue

## Testing

- Tested with incoming public reply activities to ensure they reach followers
- Verified additional recipients in the addressing fields are still processed
- Confirmed private/direct activities are unaffected by the change

## Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Set up WordPress instance with ActivityPub plugin
* Follow a remote user from Actor A
* Have Actor B (not followed by Actor A) reply to a public post from the followed user
* Verify Actor A can see the reply in their inbox

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [x] Patch
- [ ] Minor
- [ ] Major

#### Type

- [ ] Added - for new features
- [ ] Changed - for changes in existing functionality
- [ ] Deprecated - for soon-to-be removed features
- [ ] Removed - for now removed features
- [x] Fixed - for any bug fixes
- [ ] Security - in case of vulnerabilities

#### Message

Route public replies to all actors to ensure conversation continuity.

</details>

pfefferle committed:
Route public replies to all actors instead of just followers

When processing incoming ActivityPub activities, public replies should be
visible to all actors, not just those who follow the reply author. This
change updates the routing logic to deliver public replies to all actors
while maintaining the existing behavior for non-reply public activities
(which continue to go to followers only).

pfefferle committed:
Merge branch 'trunk' into feature/route-public-replies-to-all-actors

pfefferle committed:
Add changelog entry for #2585

pfefferle committed:
Revert "Add changelog entry for #2585"

This reverts commit 723eabeae5ce076abf31e7603e02f973cd734cd7.

matticbot committed:
Add changelog

pfefferle committed:
Simplify recipient logic in get_local_recipients

Refactored the get_local_recipients method to remove unnecessary branching for public activity replies and streamline how user IDs are collected for public activities.

obenland commented on includes/rest/class-inbox-controller.php:
```diff
@@ -368,8 +369,12 @@ public function get_item_schema() {
+				return Actors::get_all_ids();
+			} else {
+				return Following::get_follower_ids( $activity['actor'] );
+			}
```
```suggestion
		$user_ids = array();
		if ( is_activity_public( $activity ) ) {
			$user_ids = Following::get_follower_ids( $activity['actor'] );
		}
```
And then removing the `$user_ids   = array();` in line 383. I'm curious if that works?

Then we'd default to followers and add any specific account mentioned in replies.

obenland replied:
oh, done in 50cd03d.

pfefferle committed:
Add unit tests for optimized get_local_recipients method

Tests verify that:
- Public activities deliver to followers AND explicit recipients
- Deduplication works when user is both follower and explicit recipient
- Non-public activities ignore followers and only use explicit recipients

obenland reviewed (approved):
Nice
